src/box/gc.h

```diff
@@ -332,51 +348,75 @@ void
 gc_unref_checkpoint(struct gc_checkpoint_ref *ref);
 
 /**
- * Register a consumer.
+ * Register a consumer or update existing with the same uuid.
  *
  * This will stop garbage collection of WAL files newer than
  * @vclock until the consumer is unregistered or advanced.
  * @format... specifies a human-readable name of the consumer,
  * it will be used for listing the consumer in box.info.gc().
- *
- * Returns a pointer to the new consumer object or NULL on
- * memory allocation failure.
+ * Registered consumer is associated with object by @uuid, also
+ * all consumers are indexed by it internally - registered consumers
+ * can be manipulated only by uuid.
+ */
+CFORMAT(printf, 3, 4)
+void
+gc_consumer_register(const struct tt_uuid *uuid, const struct vclock *vclock,
+		     const char *format, ...);
```

Comment:
This change makes the API confusing and inconsistent.

Can't we leave the old API as is and instead add a couple of new functions? Here's how I see it:
1. GC fiber writes all registered consumers to the system space in the background.
2. On recovery, all consumers are loaded and marked as "orphan".
3. `gc_consumer_register` adopts the consumer from the orphan list if it finds it there.
4. `gc_consumer_unregister` leaves the consumer in the orphan list unless it's "forgotten" (see p.6).
5. `gc_sync` (new function) waits until GC fiber writes all consumers to the system space.
6. `gc_forget` (new function; may need a better name) synchronously deletes the given consumer (by id) from the system space and marks the in-memory consumer (it it exists) so that it's deleted immediately on `gc_consumer_unregister` or deletes it right away if it's unused.

Then we'd just need to put `gc_sync` after join/subscribe and `gc_forget` to `_cluster.on_replace` trigger. No global refactoring would be required - all the changes would be concentrated in `src/box/gc.c`.